google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,536 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_RESUMABLE_ASYNC_STREAMING_READ_WRITE_RPC_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_RESUMABLE_ASYNC_STREAMING_READ_WRITE_RPC_H
+
+#include "google/cloud/pubsublite/internal/futures.h"
+#include "google/cloud/async_streaming_read_write_rpc.h"
+#include "google/cloud/internal/backoff_policy.h"
+#include "google/cloud/internal/retry_policy.h"
+#include "google/cloud/log.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <memory>
+#include <utility>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+using google::cloud::internal::BackoffPolicy;
+using google::cloud::internal::RetryPolicy;
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to create new streams.
+ */
+template <typename RequestType, typename ResponseType>
+using AsyncStreamFactory = std::function<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>()>;
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to initialize a stream
+ * from AsyncStreamFactory.
+ */
+template <typename RequestType, typename ResponseType>
+using StreamInitializer = std::function<future<StatusOr<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>(
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>)>;
+
+using AsyncSleeper = std::function<future<void>(std::chrono::duration<double>)>;
+
+using RetryPolicyFactory = std::function<std::unique_ptr<RetryPolicy>()>;
+
+template <typename RequestType, typename ResponseType>
```

Comment:
Please describe what is this interface *for*.  Why would a library implementor want to have one of these around?  What responsibilities do concrete version of this class have?

Maybe add some small example showing how to use this class, even if the example uses blocking calls.  Or use fancy C++20 coroutines, e.g. (I may be using your class wrong, show me how to use it right instead):

```cc
future<Status> F(std::unique_ptr<ResumableAsyncStreamingReadWriteRpc<Req, Res>> stream) {
  auto start = co_await stream->Start();
  if (!start.ok()) co_return start;

  for (int i = 0; i != 10; ++i) {
    auto write = co_await stream->Write(MakeRequest(), grpc::WriteOptions());
    if (!write) break;
    auto read = co_await stream->Read();
    if (!read.has_value()) break;
    std::cout << *read << "\n";
  }
  co_await stream->Finish();
  co_return start;
}
```

Is that what we have in mind?  And presumably any transient errors in the `for()` loop are handled by the class, the connection is re-established and so on?  This does not seem usable to me, because the caller has no way to extract the information about any unrecoverable errors that causes the retry loop to be exhausted.

Or maybe you intended the class to be used like below, where the `start` future is the last thing returned?

```cc
future<Status> F(std::unique_ptr<ResumableAsyncStreamingReadWriteRpc<Req, Res>> stream) {
  auto start = stream->Start();

  for (int i = 0; i != 10; ++i) {
    auto write = co_await stream->Write(MakeRequest(), grpc::WriteOptions());
    if (!write) break;
    auto read = co_await stream->Read();
    if (!read.has_value()) break;
    std::cout << *read << "\n";
  }
  co_await stream->Finish();
  co_return co_await start;
}
```

If so, how do I know when is it Okay to start calling `Read()` and `Write()`?

Reply:
https://github.com/googleapis/google-cloud-cpp/pull/8441